boto3/s3/inject.py

```diff
@@ -356,3 +367,267 @@ def object_copy(self, CopySource, ExtraArgs=None, Callback=None,
         CopySource=CopySource, Bucket=self.bucket_name, Key=self.key,
         ExtraArgs=ExtraArgs, Callback=Callback, SourceClient=SourceClient,
         Config=Config)
+
+
+def upload_fileobj(self, Fileobj, Bucket, Key, ExtraArgs=None,
+                   Callback=None, Config=None):
+    """Upload a file-like object to S3.
+
+    This is a managed transfer which will perform a multipart upload in
+    multiple threads if necessary.
+
+    Usage::
+
+        import io
```

Comment:
It might be worth avoiding `io` for now, because it would be nice to have an equivalent of `boto.set_contents_from_string` which would probably be the preferred way to upload bytes-like objects instead of having to wrap it in an `io.BytesIO` object.